core/server/policy_violations.go

```diff
@@ -0,0 +1,263 @@
+	clist := clustersmngr.NewClusteredList(func() sigsClient.ObjectList {
+		return &v1.EventList{}
+	})
+
+	if err := clusterClient.ClusteredList(ctx, clist, true, opts...); err != nil {
+		var errs clustersmngr.ClusteredListError
+		if !errors.As(err, &errs) {
+			return nil, fmt.Errorf("error while listing events: %w", err)
+		}
+
+		for _, e := range errs.Errors {
+			respErrors = append(respErrors, &pb.ListError{ClusterName: e.Cluster, Message: e.Err.Error()})
+		}
+	}
+
+	var validations []*pb.PolicyValidation
+	for listClusterName, lists := range clist.Lists() {
+		if clusterName != "" && listClusterName != clusterName {
+			continue
+		}
+		for _, l := range lists {
+			list, ok := l.(*v1.EventList)
+			if !ok {
+				continue
+			}
+			for i := range list.Items {
+				validation, err := toPolicyValidation(list.Items[i], listClusterName, extraDetails)
+				if err != nil {
+					return nil, fmt.Errorf("error while getting policy violation event details: %w", err)
+				}
+				validations = append(validations, validation)
+			}
+		}
+	}
+
+	return &validationList{
+		Validations: validations,
+		Token:       clist.GetContinue(),
+		Errors:      respErrors,
+	}, nil
+}
+
+func toPolicyValidation(item v1.Event, clusterName string, extraDetails bool) (*pb.PolicyValidation, error) {
+	annotations := item.GetAnnotations()
+	policyValidation := &pb.PolicyValidation{
+		Id:          getAnnotation(item.GetLabels(), "pac.weave.works/id"),
+		Name:        getAnnotation(annotations, "policy_name"),
+		PolicyId:    getAnnotation(annotations, "policy_id"),
+		ClusterId:   getAnnotation(annotations, "cluster_id"),
+		Category:    getAnnotation(annotations, "category"),
+		Severity:    getAnnotation(annotations, "severity"),
+		CreatedAt:   item.GetCreationTimestamp().Format(time.RFC3339),
+		Message:     item.Message,
+		Entity:      item.InvolvedObject.Name,
+		Namespace:   item.InvolvedObject.Namespace,
+		ClusterName: clusterName,
+	}
+	if extraDetails {
+		policyValidation.Description = getAnnotation(annotations, "description")
+		policyValidation.HowToSolve = getAnnotation(annotations, "how_to_solve")
+		policyValidation.ViolatingEntity = getAnnotation(annotations, "entity_manifest")
+		err := json.Unmarshal([]byte(getAnnotation(annotations, "occurrences")), &policyValidation.Occurrences)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get occurrences from event: %w", err)
+		}
+		paramsRaw := getAnnotation(annotations, "parameters")
+		if paramsRaw != "" {
+			parameter, err := getPolicyValidationParam([]byte(paramsRaw))
+			if err != nil {
+				return nil, err
+			}
+			policyValidation.Parameters = parameter
+		}
+	}
+	return policyValidation, nil
+}
+
+func getAnnotation(annotations map[string]string, key string) string {
+	value, ok := annotations[key]
+	if !ok {
+		return ""
+	}
+	return value
+}
+
+func getPolicyValidationParam(raw []byte) ([]*pb.PolicyValidationParam, error) {
+	var paramsArr []map[string]interface{}
```

Comment:
Can this be described by a `struct` instead? I see that we have known fields below, which makes me thing we could do:

```go
type validationParam struct {
  Name string `json:"name"`
  ...
}
```

Reply:
We already have `PolicyValidationParam` struct for the response but the `value` field could have different types depending on the policy (could be numeric or boolean or string) it's already implemented as `google.protobuf.Any` in the proto response